src/ar_tx_db.erl

```diff
@@ -15,36 +20,58 @@ start() ->
 			receive stop -> ok end
 		end
 	),
-	% Add a short wait to ensure that the table has been created
-	% before returning.
+	%% Add a short wait to ensure that the table has been created
+	%% before returning.
 	receive after 250 -> ok end.
 
-%% @doc Put an Erlang term into the meta DB. Typically these are
-%% write-once values.
 put_error_codes(TXID, ErrorCodes) ->
-	ets:insert(?MODULE, {TXID, ErrorCodes}),
-	{ok, _} = timer:apply_after(1800*1000, ?MODULE, clear_error_codes, [TXID]),
-	ok.
+	log(TXID, {validation_error, ErrorCodes}).
 
-%% @doc Retreive a term from the meta db.
 get_error_codes(TXID) ->
-	case ets:lookup(?MODULE, TXID) of
-		[{_, ErrorCodes}] -> {ok, ErrorCodes};
-		[] -> not_found
-	end.
+	Log = get_tx_log(TXID),
+	search_for_validation_error(Log).
+
+search_for_validation_error([{{validation_error, ErrorCodes}, _} | _] ) ->
```

Comment:
If this is an ETS table, couldn't you use a match or a select to go through the records? Or is it not worth the optimisation?

Reply:
`match` makes sense 👍

Reply:
On second thought, matching is helpful if the key is specified in the pattern. In this case the event list is fetched by TXID first, and further searching inside the list is not something `ets:match` can help with.